zebra-chain/src/sprout/tree.rs

```diff
@@ -116,186 +128,181 @@ impl From<&Root> for [u8; 32] {
     }
 }
 
-/// Sprout Note Commitment Tree
-#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
-#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+/// A node of the Sprout note commitment tree.
+#[derive(Clone, Debug)]
+struct Node([u8; 32]);
+
+impl incrementalmerkletree::Hashable for Node {
+    /// Returns an empty leaf.
+    fn empty_leaf() -> Self {
+        Self(NoteCommitmentTree::uncommitted())
+    }
+
+    /// Combines two nodes to generate a new node.
+    ///
+    /// Note that Sprout does not use the `level` argument.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#sproutmerklecrh
+    fn combine(_level: incrementalmerkletree::Altitude, a: &Self, b: &Self) -> Self {
+        Self(merkle_crh_sprout(a.0, b.0))
+    }
+
+    /// Returns the node for the level below the given level. (A quirk of the API)
+    fn empty_root(level: incrementalmerkletree::Altitude) -> Self {
+        let layer_below: usize = MERKLE_DEPTH - usize::from(level);
+        Self(EMPTY_ROOTS[layer_below])
+    }
+}
+
+impl From<NoteCommitment> for Node {
+    fn from(cm: NoteCommitment) -> Self {
+        Node(cm.into())
+    }
+}
+
+impl serde::Serialize for Node {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for Node {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let bytes = <[u8; 32]>::deserialize(deserializer)?;
+        let cm = NoteCommitment::from(bytes);
+        let node = Node::from(cm);
+
+        Ok(node)
+    }
+}
+
+#[allow(dead_code, missing_docs)]
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+pub enum NoteCommitmentTreeError {
+    #[error("the note commitment tree is full")]
+    FullTree,
+}
+
+/// Sprout Note Commitment Tree.
+///
+/// The tree is implemented as an incremental Merkle tree.
```

Comment:
```suggestion
/// [Sprout Note Commitment Tree].
///
/// An incremental Merkle tree of fixed depth used to store Sprout note commitments. 
/// It is used to express the existence of value and the capability to spend it. It is _not_ the 
/// job of this tree to protect against double-spending, as it is append-only; double-spending 
/// is prevented by maintaining the [nullifier set] for each shielded pool.
///
/// Internally this wraps [`incrementalmerkletree::Frontier`], so that we can maintain and increment
/// the full tree with only the minimal amount of non-empty nodes/leaves required.
/// 
/// [Sprout Note Commitment Tree]: https://zips.z.cash/protocol/protocol.pdf#merkletree
/// [nullifier set]: https://zips.z.cash/protocol/protocol.pdf#nullifierset
```